File: app/lib/vc.ts
/**
 * The personalisation client: session, VC profile, survey, decisions, gap, personal rank.
 *
 * WHY THIS IS NOT `lib/api.ts`. That module is fallback-first: when the backend is
 * unreachable it serves hand-authored fixtures so the objective product still renders.
 * That contract is exactly wrong here. A fixture profile is an INVENTED profile, and the
 * one rule this feature cannot break is that everything on screen about the user is
 * something the user actually submitted (docs/DIFFERENTIATOR.md §2.2 — a profile
 * inferred from four decisions must say so). So this module has no fixtures at all: a
 * failed call returns `{ ok: false, error }` and the page renders the failure, never a
 * plausible persona.
 *
 * What it shares with `lib/api.ts` is the transport discipline: NOTHING HERE REJECTS.
 * Every function resolves to a discriminated result, so no caller can be left with a
 * disabled button and nothing on screen to explain it.
 *
 * Two transport rules, both load-bearing:
 *   - `credentials: "include"` on every call. The session is an httpOnly cookie, so it
 *     is invisible to JS and travels only if the fetch asks for it. One missing flag
 *     here makes every authenticated route return 401 while login looks like it worked.
 *   - A 401 is a STATE, not an error. It means "anonymous", which is a legitimate,
 *     fully-rendering path (§1). Callers distinguish it via `status`.
 */

import { API_BASE, TIMEOUT } from "./api";

export type Res<T> =
  | { ok: true; data: T }
  | { ok: false; error: string; status?: number };

// ---------------------------------------------------------------------------
// Wire types. Mirrors of schema/vc.py — kept narrow to what the UI renders.
// ---------------------------------------------------------------------------

export interface Provenance {
  basis: string;
  method: string;
  question_ids: string[];
  decision_rows: number[];
  n: number;
}

export interface NotInferred {
  field_name: string;
  reason: string;
}

export interface User {
  user_id: string;
  email: string;
  created_at: string;
}

export interface Me {
  authenticated: boolean;
  user: User | null;
  personalisation_enabled: boolean;
  reason: string;
}

export interface SurveyOption {
  text: string;
  signals: Record<string, number>;
}

export interface SurveyQuestion {
  id: string;
  prompt: string;
  option_a: SurveyOption;
  option_b: SurveyOption;
}

export interface Survey {
  questions: SurveyQuestion[];
  /** question_id -> "a" | "b". Absent means UNANSWERED, which is not the same as neutral. */
  answers: Record<string, string>;
  answered: number;
  total: number;
}

export interface AxisWeights {
  founder: number;
  market: number;
  idea_vs_market: number;
  provenance: Provenance;
  confidence: number;
}

export interface ConvictionStyle {
  score: number;
  label: string;
  provenance: Provenance;
  confidence: number;
}

export interface Prior {
  key: string;
  count: number;
  share: number;
  provenance: Provenance;
}

export interface RedLine {
  statement: string;
  source: string;
  provenance: Provenance;
  confidence: number;
}

export interface Derived {
  axis_weights_stated: AxisWeights | null;
  axis_weights_revealed: AxisWeights | null;
  conviction_style_stated: ConvictionStyle | null;
  conviction_style_revealed: ConvictionStyle | null;
  sector_priors: Prior[];
  stage_priors: Prior[];
  red_lines: RedLine[];
  survey_answered: number;
  survey_total: number;
  decisions_count: number;
  invested_count: number;
  confidence: number;
  personalisation_enabled: boolean;
  personalisation_reason: string;
  not_inferred: NotInferred[];
}

export interface Profile {
  profile_id: string;
  user_id: string;
  fund_name: string | null;
  focus_sectors: string[];
  stated_red_lines: string[];
  updated_at: string;
  derived: Derived;
}

export interface RejectedRow {
  row_number: number;
  reason: string;
  raw: string;
}

export interface UploadResult {
  accepted: number;
  rejected: RejectedRow[];
  warnings: RejectedRow[];
  total_rows: number;
}

export interface GapFinding {
  dimension: string;
  stated: string;
  revealed: string;
  finding: string;
  magnitude: number;
  provenance: Provenance;
  confidence: number;
}

export interface GapUncomputable {
  dimension: string;
  missing: string;
  reason: string;
}

export interface GapReport {
  findings: GapFinding[];
  uncomputable: GapUncomputable[];
  agreements: string[];
  computed_at: string;
  personalisation_enabled: boolean;
  personalisation_reason: string;
}

export interface Lens {
  kind: string;
  persona: string;
  weight: number;
  justified_by: string[];
  provenance: Provenance;
  confidence: number;
}

export interface PersonalRankRow {
  company_id: string;
  name: string;
  core_rank: number;
  personal_rank: number;
  fit_score: number;
  core_weakest_score: number;
  /** Positive = the personal layer PROMOTED this against core. */
  divergence: number;
  top_lens: string | null;
  why: string;
}

export interface Disagreement {
  company_id: string;
  name: string;
  core_rank: number;
  personal_rank: number;
  divergence: number;
  explanation: string;
}

export interface CoreRankRow {
  company_id: string;
  name: string;
  core_rank: number;
}

export interface PersonalRanking {
  as_of: string;
  personalised: boolean;
  reason: string;
  lenses: Lens[];
  lenses_not_derived: NotInferred[];
  rows: PersonalRankRow[];
  disagreements: Disagreement[];
  agreements: string[];
  /** Served unconditionally, including when personalisation is off (§3). */
  core_rank: CoreRankRow[];
}

// ---------------------------------------------------------------------------
// Transport
// ---------------------------------------------------------------------------

/**
 * One request. Resolves, always.
 *
 * FastAPI puts the human-readable cause in `detail`, and on the personal routes that
 * `detail` is sometimes an object carrying the reason personalisation is off — which is
 * the most useful sentence on the screen, so it is unwrapped rather than stringified
 * into "[object Object]".
 */
async function call<T>(
  path: string,
  init: RequestInit & { timeoutMs?: number } = {},
): Promise<Res<T>> {
  const { timeoutMs = TIMEOUT.read, ...rest } = init;
  const ctrl = new AbortController();
  let timedOut = false;
  const timer = setTimeout(() => {
    timedOut = true;
    ctrl.abort();
  }, timeoutMs);
  try {
    const res = await fetch(`${API_BASE}${path}`, {
      ...rest,
      signal: ctrl.signal,
      cache: "no-store",
      // The whole feature depends on this line.
      credentials: "include",
      headers: { accept: "application/json", ...(rest.headers ?? {}) },
    });
    const body: unknown = await res.json().catch(() => null);
    if (!res.ok) {
      return { ok: false, error: detailOf(body) ?? `${res.status} ${res.statusText}`, status: res.status };
    }
    return { ok: true, data: body as T };
  } catch (e) {
    return {
      ok: false,
      error: timedOut
        ? `no response in ${timeoutMs / 1000}s`
        : e instanceof Error
          ? e.message
          : String(e),
    };
  } finally {
    clearTimeout(timer);
  }
}

function detailOf(body: unknown): string | null {
  if (!body || typeof body !== "object") return null;
  const detail = (body as { detail?: unknown }).detail;
  if (typeof detail === "string") return detail;
  if (detail && typeof detail === "object") {
    const d = detail as { error?: unknown; reason?: unknown };
    const parts = [d.error, d.reason].filter((p): p is string => typeof p === "string");
    if (parts.length) return parts.join(" — ");
  }
  return null;
}

const json = (body: unknown): RequestInit => ({
  method: "POST",
  headers: { "content-type": "application/json" },
  body: JSON.stringify(body),
});

// ---------------------------------------------------------------------------
// Auth
// ---------------------------------------------------------------------------

/** The anonymous state, rendered when `/auth/me` cannot be reached at all.
 *
 *  This is NOT a fixture in the `lib/api.ts` sense — it invents nothing about a user.
 *  It is the honest reading of "we could not establish a session", and it is what keeps
 *  a dead auth service from taking the core product down with it (§1). */
export function anonymous(reason: string): Me {
  return { authenticated: false, user: null, personalisation_enabled: false, reason };
}

export async function getMe(): Promise<Me> {
  const r = await call<Me>("/auth/me");
  if (r.ok && typeof r.data?.authenticated === "boolean") return r.data;
  return anonymous(
    r.ok
      ? "/auth/me returned a shape this client cannot read — treating the session as absent"
      : `the session could not be checked (${r.error}) — the core objective ranking is unaffected`,
  );
}

/**
 * Register. The API deliberately DOES disclose that an email is taken here (see
 * `api/routers/auth.py`), because the alternative is telling a returning user "success"
 * and then failing their login. Login is the route that must not disclose, and does not.
 */
export function register(
  email: string,
  password: string,
  fundName?: string,
): Promise<Res<{ user: User }>> {
  return call<{ user: User }>(
    "/auth/register",
    json({ email, password, fund_name: fundName?.trim() || null }),
  );
}

export function login(email: string, password: string): Promise<Res<{ user: User }>> {
  return call<{ user: User }>("/auth/login", json({ email, password }));
}

export function logout(): Promise<Res<unknown>> {
  return call<unknown>("/auth/logout", { method: "POST" });
}

// ---------------------------------------------------------------------------
// Profile
// ---------------------------------------------------------------------------

export const getProfile = () => call<Profile>("/profile");

export const putProfile = (body: {
  fund_name?: string | null;
  focus_sectors?: string[];
  stated_red_lines?: string[];
}) => call<Profile>("/profile", { ...json(body), method: "PUT" });

export const getSurvey = () => call<Survey>("/profile/survey");

export const postSurvey = (answers: { question_id: string; choice: string }[]) =>
  call<{ stored: number; answered: number; total: number; derived: Derived }>(
    "/profile/survey",
    json({ answers }),
  );

/** The upload posts the file's text as the raw body; the API detects CSV vs JSON from
 *  the CONTENT, so the extension is never treated as a fact. */
export const postDecisions = (content: string) =>
  call<{ upload: UploadResult; derived: Derived }>("/profile/decisions", {
    method: "POST",
    headers: { "content-type": "text/plain" },
    body: content,
    timeoutMs: TIMEOUT.query,
  });

export const getGap = () => call<GapReport>("/profile/gap");

// ---------------------------------------------------------------------------
// Personal layer
// ---------------------------------------------------------------------------

/**
 * The LLM budget, not the query budget. This endpoint replays all 113 companies
 * through the composed council and measured 10.8s against the deployed backend
 * on a WARM instance — a cold start pushes it past the 15s query budget, at
 * which point the pipeline silently rendered "—" for every YOUR RANK while
 * personalisation was on. The caller treats a failure as "no personal rank",
 * so an aborted request is indistinguishable from a signed-out visitor; the
 * budget has to clear the slow case for the distinction to survive.
 */
export const getPersonalRank = () =>
  call<PersonalRanking>("/personal/rank", { timeoutMs: TIMEOUT.llm });

/** One stored authored agent, exactly as `AuthoredLens` serialises on the wire. */
export interface AuthoredLensRecord {
  lens_id: string;
  name: string;
  quality: string;
  persona: string;
  weight: number;
  origin: "authored" | "template";
  created_at: string;
  updated_at: string;
}

export interface LensSet {
  personalisation_enabled: boolean;
  personalisation_reason: string;
  profile_confidence: number;
  /** The DERIVED half only, weighted among themselves — the old contract, unchanged. */
  lenses: Lens[];
  not_derived: NotInferred[];
  /** The stored records the VC owns and edits. */
  authored: AuthoredLensRecord[];
  /** The council that actually scores: derived + authored at the ranking's weights. */
  council: Lens[];
  council_not_derived: NotInferred[];
  weight_rule: string;
  refusal: { reason: string } | null;
  min_lenses: number;
  max_lenses: number;
  sufficient: boolean;
  authored_survive_rederive: boolean;
}

/** A create body for one authored agent — mirrors `AuthoredLensWrite`. */
export interface LensWrite {
  name: string;
  quality: string;
  persona: string;
  weight: number;
  origin: "authored" | "template";
}

/**
 * This VC's council: derived, authored, and the composition the ranking uses.
 * Every write route below returns this same payload, so a create and a delete hand
 * back the council the GET does and no client reconciles two views of it.
 */
export const getLenses = () => call<LensSet>("/personal/lenses");

/**
 * Council writes. The server is the authority on every bound: the 422 for a quality
 * with no readable term, the 409 at the agent ceiling. Callers surface those reasons
 * verbatim instead of pre-empting them — the refusal text is the feature.
 * `TIMEOUT.query` rather than `read`: a write that lands on a cold function should
 * survive the import, not report a failure for a lens that was in fact created.
 */
export const createLens = (body: LensWrite) =>
  call<LensSet & { created: AuthoredLensRecord }>("/personal/lenses", {
    ...json(body),
    timeoutMs: TIMEOUT.query,
  });

export const updateLens = (lensId: string, patch: Partial<LensWrite>) =>
  call<LensSet & { updated: AuthoredLensRecord }>(
    `/personal/lenses/${encodeURIComponent(lensId)}`,
    // `json()` bakes in POST, so the method override must come after the spread.
    { ...json(patch), method: "PUT", timeoutMs: TIMEOUT.query },
  );

export const deleteLens = (lensId: string) =>
  call<LensSet & { deleted: string }>(`/personal/lenses/${encodeURIComponent(lensId)}`, {
    method: "DELETE",
    timeoutMs: TIMEOUT.query,
  });

File: schema/vc.py
"""The VC-side contract: accounts, the two preference sources, and what we derive.

Owner: personalisation layer (docs/DIFFERENTIATOR.md §1-§2).

THE LOAD-BEARING RULE (§0): stated preference (survey) and revealed preference (past
decisions) are two SEPARATE inputs and are never merged into one blob. They are stored
in separate tables, derived independently, and compared in `GapReport`. Merging them
would save code and destroy the only finding in here that is about the USER rather than
about a founder.

THE SECOND RULE: nothing in a derived profile may exist without a real submission behind
it. Every derived value carries `Provenance` naming the exact questions or decision rows
that produced it, plus a confidence. Where the submitted data cannot support an
inference, the field is ABSENT and the reason is recorded in `DerivedProfile.not_inferred`
— never filled with a plausible default. This is the same discipline the memo already
follows: gaps are flagged, never filled.
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Optional
from uuid import UUID, uuid4

from pydantic import BaseModel, Field, field_validator

from ._backport import StrEnum


def normalize_email(raw: str) -> str:
    """Lowercased and trimmed, with a deliberately minimal shape check.

    Not `EmailStr`: that pulls in `email-validator`, and the Vercel bundle is already
    against a 250MB ceiling. The address is an account key here, never something we
    send mail to, so shape is all that has to hold.
    """
    email = (raw or "").strip().lower()
    local, _, domain = email.partition("@")
    if not local or not domain or "." not in domain or " " in email:
        raise ValueError("not a valid email address")
    return email

# ---------------------------------------------------------------------------
# Accounts
# ---------------------------------------------------------------------------


class User(BaseModel):
    """An account. `password_hash` is deliberately NOT on this model — it lives only in
    the storage layer, so there is no route that can accidentally serialise it."""

    user_id: UUID = Field(default_factory=uuid4)
    email: str
    created_at: datetime


class RegisterRequest(BaseModel):
    email: str
    # A floor, not a composition rule. Length is the only password requirement with
    # evidence behind it; the upper bound stops a multi-megabyte body from becoming
    # an argon2 denial-of-service.
    password: str = Field(min_length=10, max_length=1024)
    fund_name: Optional[str] = None

    @field_validator("email")
    @classmethod
    def _email(cls, v: str) -> str:
        return normalize_email(v)


class LoginRequest(BaseModel):
    email: str
    password: str = Field(max_length=1024)

    @field_validator("email")
    @classmethod
    def _email(cls, v: str) -> str:
        return normalize_email(v)


# ---------------------------------------------------------------------------
# Input 1: the survey — STATED preference
# ---------------------------------------------------------------------------

# Signal keys an option may emit. Constrained, because a free-form key would silently
# vanish from the derivation instead of failing loudly.
#   founder / market / idea_vs_market -> the three axes the core screen already uses
#   conviction   -1 = evidence-heavy, +1 = conviction-heavy
#   stage_lean   -1 = earliest stage,  +1 = later stage
AXIS_SIGNALS = ("founder", "market", "idea_vs_market")
SCALAR_SIGNALS = ("conviction", "stage_lean")
SIGNAL_KEYS = AXIS_SIGNALS + SCALAR_SIGNALS


class Choice(StrEnum):
    A = "a"
    B = "b"


class SurveyOption(BaseModel):
    text: str
    signals: dict[str, float] = Field(default_factory=dict)

    @field_validator("signals")
    @classmethod
    def _known_keys(cls, v: dict[str, float]) -> dict[str, float]:
        unknown = set(v) - set(SIGNAL_KEYS)
        if unknown:
            raise ValueError(f"unknown signal keys: {sorted(unknown)}")
        return v


class SurveyQuestion(BaseModel):
    """A forced trade-off. Not a Likert item, on purpose: agreement scales measure
    agreeableness, trade-offs measure priorities (§2.1)."""

    id: str
    prompt: str
    option_a: SurveyOption
    option_b: SurveyOption

    def option(self, choice: Choice) -> SurveyOption:
        return self.option_a if choice == Choice.A else self.option_b


class SurveyAnswer(BaseModel):
    question_id: str
    choice: Choice


class SurveySubmission(BaseModel):
    """Partial submissions are legal. An unanswered question contributes NOTHING and
    lowers the confidence — it is never imputed from the answers around it."""

    answers: list[SurveyAnswer] = Field(default_factory=list)


# THE SURVEY. Twelve forced trade-offs, each one a choice a working investor actually
# has to make. Every option carries the signals it implies; a question that implies
# nothing about a dimension emits nothing for it rather than a token weight.
SURVEY_QUESTIONS: list[SurveyQuestion] = [
    SurveyQuestion(
        id="q01_founder_vs_market",
        prompt=(
            "A technically exceptional founder in a crowded market, or a competent "
            "operator in an empty one."
        ),
        option_a=SurveyOption(text="Exceptional founder, crowded market", signals={"founder": 1.0}),
        option_b=SurveyOption(text="Competent operator, empty market", signals={"market": 1.0}),
    ),
    SurveyQuestion(
        id="q02_traction_vs_demo",
        prompt=(
            "Eighteen months of real but flat usage, or a six-week-old prototype that "
            "makes you feel something."
        ),
        option_a=SurveyOption(
            text="Eighteen months of flat real usage",
            signals={"conviction": -1.0, "stage_lean": 1.0, "market": 0.5},
        ),
        option_b=SurveyOption(
            text="Six-week-old prototype with a pull",
            signals={"conviction": 1.0, "stage_lean": -1.0, "idea_vs_market": 0.5},
        ),
    ),
    SurveyQuestion(
        id="q03_early_vs_ontime",
        prompt="The right product two years early, or the second-best product exactly on time.",
        option_a=SurveyOption(
            text="Right product, two years early",
            signals={"idea_vs_market": 1.0, "conviction": 0.5},
        ),
        option_b=SurveyOption(
            text="Second-best product, perfect timing",
            signals={"market": 1.0, "conviction": -0.3},
        ),
    ),
    SurveyQuestion(
        id="q04_velocity_vs_durability",
        prompt=(
            "A solo founder shipping every week, or a two-person team shipping monthly "
            "whose skills genuinely complement each other."
        ),
        option_a=SurveyOption(
            text="Solo founder, weekly shipping", signals={"founder": 1.0, "conviction": 0.5}
        ),
        option_b=SurveyOption(
            text="Complementary pair, monthly shipping",
            signals={"founder": 0.5, "conviction": -0.3},
        ),
    ),
    SurveyQuestion(
        id="q05_distribution_vs_product",
        prompt=(
            "An average product with a founder who can sell anything, or an extraordinary "
            "product built by a team that has never spoken to a customer."
        ),
        option_a=SurveyOption(
            text="Average product, exceptional seller", signals={"market": 1.0, "founder": 0.5}
        ),
        option_b=SurveyOption(
            text="Exceptional product, no customer contact", signals={"idea_vs_market": 1.0}
        ),
    ),
    SurveyQuestion(
        id="q06_which_mistake",
        prompt=(
            "You will be wrong either way. Would you rather miss a fund-returner, or "
            "write a cheque that goes to zero?"
        ),
        option_a=SurveyOption(
            text="Rather miss the fund-returner", signals={"conviction": -1.0, "stage_lean": 1.0}
        ),
        option_b=SurveyOption(
            text="Rather write the cheque that dies", signals={"conviction": 1.0, "stage_lean": -1.0}
        ),
    ),
    SurveyQuestion(
        id="q07_metrics_vs_love",
        prompt=(
            "Unverifiable numbers but customers who will not shut up about it, or clean "
            "audited metrics and customers who are merely satisfied."
        ),
        option_a=SurveyOption(
            text="Unverifiable numbers, fanatical customers",
            signals={"conviction": 1.0, "market": 0.5, "stage_lean": -0.5},
        ),
        option_b=SurveyOption(
            text="Clean metrics, indifferent customers",
            signals={"conviction": -1.0, "stage_lean": 0.5},
        ),
    ),
    SurveyQuestion(
        id="q08_incumbent_vs_demand",
        prompt=(
            "The main risk is that the incumbent ships it next quarter, or the main risk "
            "is that nobody turns out to want it."
        ),
        option_a=SurveyOption(
            text="Can live with incumbent risk", signals={"idea_vs_market": 1.0, "founder": 0.5}
        ),
        option_b=SurveyOption(text="Can live with demand risk", signals={"market": 1.0}),
    ),
    SurveyQuestion(
        id="q09_adaptability_vs_persistence",
        prompt=(
            "A founder who has changed the plan three times this year, or one who has held "
            "the same plan for three years."
        ),
        option_a=SurveyOption(
            text="Changed the plan three times", signals={"founder": 0.5, "conviction": 0.5}
        ),
        option_b=SurveyOption(
            text="Held the plan for three years", signals={"founder": 0.5, "conviction": -0.5}
        ),
    ),
    SurveyQuestion(
        id="q10_insider_vs_outsider",
        prompt=(
            "A domain expert who has lived the problem for a decade, or an outsider who "
            "finds the entire industry's assumptions absurd."
        ),
        option_a=SurveyOption(text="Ten-year domain insider", signals={"founder": 1.0}),
        option_b=SurveyOption(
            text="Outsider who rejects the premises",
            signals={"idea_vs_market": 1.0, "conviction": 0.5},
        ),
    ),
    SurveyQuestion(
        id="q11_speed_vs_diligence",
        prompt=(
            "Decide in one meeting and be occasionally reckless, or take six weeks of "
            "diligence and occasionally lose the deal."
        ),
        option_a=SurveyOption(
            text="One meeting, occasionally reckless",
            signals={"conviction": 1.0, "stage_lean": -0.5},
        ),
        option_b=SurveyOption(
            text="Six weeks, occasionally too slow",
            signals={"conviction": -1.0, "stage_lean": 0.5},
        ),
    ),
    SurveyQuestion(
        id="q12_price_vs_pick",
        prompt=(
            "Your highest-conviction company at a price you think is twice what it should "
            "be, or your third-favourite at a price you think is fair."
        ),
        option_a=SurveyOption(
            text="Best company, bad price", signals={"conviction": 1.0, "founder": 0.5}
        ),
        option_b=SurveyOption(text="Third-best company, fair price", signals={"conviction": -1.0}),
    ),
]

SURVEY_BY_ID = {q.id: q for q in SURVEY_QUESTIONS}


# ---------------------------------------------------------------------------
# Input 2: past decisions — REVEALED preference
# ---------------------------------------------------------------------------


class DecisionKind(StrEnum):
    INVESTED = "invested"
    PASSED = "passed"
    WATCHED = "watched"


class PastDecision(BaseModel):
    company: str
    sector: Optional[str] = None
    stage: Optional[str] = None
    decision: DecisionKind
    decided_on: Optional[date] = None
    rationale: Optional[str] = None
    outcome: Optional[str] = None
    source_row: Optional[int] = None  # 1-based row in the uploaded file: the provenance handle


class RejectedRow(BaseModel):
    """A row we could not read. Reported, never silently dropped — a parser that
    quietly discards a third of an upload produces a confident profile of nothing."""

    row_number: int
    reason: str
    raw: str


class DecisionUploadResult(BaseModel):
    """`accepted + len(rejected) == total_rows` always holds — every row is accounted for.

    `warnings` are rows that WERE accepted but lost an optional field on the way in (an
    unreadable date, say). They are reported rather than silently degraded, but they are
    not rejections: a row whose date we cannot read still has a perfectly legible
    verdict and sector, and throwing it away would discard real revealed preference to
    protect a field no derivation reads.
    """

    accepted: int
    rejected: list[RejectedRow] = Field(default_factory=list)
    warnings: list[RejectedRow] = Field(default_factory=list)
    total_rows: int


# ---------------------------------------------------------------------------
# What we derive — every value carries where it came from
# ---------------------------------------------------------------------------


class Provenance(BaseModel):
    """Which real submissions produced a derived value. `n` is the count of underlying
    observations, and it is what drives confidence: a weight from 4 decisions has to
    say it came from 4 decisions."""

    basis: str  # "survey" | "decisions" | "profile_field"
    method: str  # one line naming the computation, in plain words
    question_ids: list[str] = Field(default_factory=list)
    decision_rows: list[int] = Field(default_factory=list)
    n: int = 0


class AxisWeights(BaseModel):
    """Relative weight on the three axes. Normalised to sum to 1 across whatever the
    submission actually supported."""

    founder: float
    market: float
    idea_vs_market: float
    provenance: Provenance
    confidence: float = Field(ge=0.0, le=1.0)


class ConvictionStyle(BaseModel):
    """-1 evidence-heavy .. +1 conviction-heavy."""

    score: float = Field(ge=-1.0, le=1.0)
    label: str
    provenance: Provenance
    confidence: float = Field(ge=0.0, le=1.0)


class Prior(BaseModel):
    """Revealed concentration in one sector or stage. `share` is over INVESTED rows
    only — a pass is not a preference for the thing passed on."""

    key: str
    count: int
    share: float
    provenance: Provenance


class RedLine(BaseModel):
    """Something disqualifying regardless of score.

    `source` is always one of:
      stated             — the user typed it. Taken at face value, confidence 1.0.
      revealed_candidate — a unanimous pass pattern over enough rows to be worth
                           showing. A CANDIDATE, never an established rule: the user
                           has to confirm it. We do not get to invent a VC's red lines.
    """

    statement: str
    source: str
    provenance: Provenance
    confidence: float = Field(ge=0.0, le=1.0)


class NotInferred(BaseModel):
    """A field we deliberately did NOT derive, and why. This model is the whole point:
    the honest answer to thin data is an empty field with a reason attached."""

    field_name: str
    reason: str


class DerivedProfile(BaseModel):
    """Derived on READ from the two raw tables — never stored as a merged blob, so the
    stated and revealed sides can always be recomputed and compared independently."""

    axis_weights_stated: Optional[AxisWeights] = None
    axis_weights_revealed: Optional[AxisWeights] = None
    conviction_style_stated: Optional[ConvictionStyle] = None
    conviction_style_revealed: Optional[ConvictionStyle] = None
    sector_priors: list[Prior] = Field(default_factory=list)
    stage_priors: list[Prior] = Field(default_factory=list)
    red_lines: list[RedLine] = Field(default_factory=list)

    survey_answered: int = 0
    survey_total: int = len(SURVEY_QUESTIONS)
    decisions_count: int = 0
    invested_count: int = 0

    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    personalisation_enabled: bool = False
    personalisation_reason: str = ""
    not_inferred: list[NotInferred] = Field(default_factory=list)


class VCProfile(BaseModel):
    profile_id: UUID
    user_id: UUID
    fund_name: Optional[str] = None
    focus_sectors: list[str] = Field(default_factory=list)
    stated_red_lines: list[str] = Field(default_factory=list)
    updated_at: datetime
    derived: DerivedProfile


class ProfileUpdate(BaseModel):
    fund_name: Optional[str] = None
    focus_sectors: Optional[list[str]] = None
    stated_red_lines: Optional[list[str]] = None


# ---------------------------------------------------------------------------
# §3 — user-authored council lenses
#
# THE DISTINCTION THIS SECTION EXISTS TO HOLD. A derived lens is an INFERENCE: the
# system read it out of survey answers or uploaded decisions, and it must name the
# profile field that justified it. An authored lens is a STATEMENT: the VC typed it,
# and there is no profile field behind it because there was never an inference. Faking
# a profile field for an authored lens would merge the stated and the derived sides of
# the profile into one blob, which is the one thing §0 forbids — the stated-vs-revealed
# gap is only computable while the two remain separable.
#
# So `LensOrigin` is a first-class field, "derived" is NOT a value any client can send,
# and the storage layer's CHECK constraint makes an authored row that claims to be
# derived impossible rather than merely discouraged.
#
# AND: nothing here is ever created implicitly. There is no default council, no seeded
# lens and no auto-accepted template. `origin` has no default precisely so that a client
# must state whether the VC typed this or knowingly accepted a template — both are real
# input, and which one it was is a fact about the user we do not get to guess.
# ---------------------------------------------------------------------------


class LensOrigin(StrEnum):
    """How a council lens came to exist.

    DERIVED is produced only by `intelligence.custom_council.derive_lenses` from profile
    fields. It is deliberately unreachable from any request body: see `AuthoredLensWrite`.
    """

    DERIVED = "derived"
    #: The VC typed this one from scratch.
    AUTHORED = "authored"
    #: The VC knowingly accepted a template and (possibly) edited it. Still real input —
    #: they chose it — but a different basis from typing it, and the record says which.
    TEMPLATE = "template"


#: The origins a user is allowed to author. "derived" is the system's word.
AUTHORABLE_ORIGINS = (LensOrigin.AUTHORED, LensOrigin.TEMPLATE)

#: An authored weight of exactly 0 is not a lens, it is a deleted lens with extra steps.
#: The council refuses it rather than storing a seat that contributes nothing.
MIN_AUTHORED_WEIGHT = 0.01


def _nonblank(value: str, field: str) -> str:
    text = (value or "").strip()
    if not text:
        raise ValueError(f"{field} cannot be blank — an unnamed council agent is not authorable")
    return text


class AuthoredLensWrite(BaseModel):
    """A create/replace body for one authored council agent.

    `origin` is required and constrained to `AUTHORABLE_ORIGINS`, so no request body can
    mint a lens that claims the system derived it.
    """

    name: str = Field(max_length=120)
    #: The quality this agent adds score for. It is not decoration: the council reads
    #: this against the company's filtered evidence graph, so a lens whose quality
    #: carries no readable term would be a seat that measures nothing.
    quality: str = Field(max_length=120)
    persona: str = Field(max_length=4000)
    weight: float = Field(ge=MIN_AUTHORED_WEIGHT, le=1.0)
    origin: LensOrigin

    @field_validator("name", "quality", "persona")
    @classmethod
    def _text(cls, v: str, info) -> str:
        return _nonblank(v, info.field_name)

    @field_validator("origin")
    @classmethod
    def _authorable(cls, v: LensOrigin) -> LensOrigin:
        if v not in AUTHORABLE_ORIGINS:
            raise ValueError(
                "origin must be 'authored' or 'template'; 'derived' is reserved for lenses "
                "the system read out of your profile and cannot be claimed by a client"
            )
        return v


class AuthoredLensPatch(BaseModel):
    """A partial edit. An omitted field is left alone rather than cleared."""

    name: Optional[str] = Field(default=None, max_length=120)
    quality: Optional[str] = Field(default=None, max_length=120)
    persona: Optional[str] = Field(default=None, max_length=4000)
    weight: Optional[float] = Field(default=None, ge=MIN_AUTHORED_WEIGHT, le=1.0)
    origin: Optional[LensOrigin] = None

    @field_validator("name", "quality", "persona")
    @classmethod
    def _text(cls, v: Optional[str], info) -> Optional[str]:
        return None if v is None else _nonblank(v, info.field_name)

    @field_validator("origin")
    @classmethod
    def _authorable(cls, v: Optional[LensOrigin]) -> Optional[LensOrigin]:
        if v is not None and v not in AUTHORABLE_ORIGINS:
            raise ValueError("origin must be 'authored' or 'template'")
        return v


class AuthoredLens(BaseModel):
    """One stored council agent, owned by a profile.

    This is MUTABLE state, like the rest of the profile and for the same reason recorded
    in migration 002: the user edits it, and an edit is not an observation about the
    world, so it does not belong in the append-only event log.
    """

    lens_id: UUID
    profile_id: UUID
    name: str
    quality: str
    persona: str
    weight: float = Field(ge=MIN_AUTHORED_WEIGHT, le=1.0)
    origin: LensOrigin
    created_at: datetime
    updated_at: datetime

    @field_validator("origin")
    @classmethod
    def _never_derived(cls, v: LensOrigin) -> LensOrigin:
        if v == LensOrigin.DERIVED:
            raise ValueError(
                "a stored authored lens can never carry origin 'derived' — it would let a "
                "lens the VC typed masquerade as one read out of their answers"
            )
        return v


# ---------------------------------------------------------------------------
# §2.3 — the stated-vs-revealed gap
# ---------------------------------------------------------------------------


class GapFinding(BaseModel):
    """A divergence between what the VC said and what the VC did. Only emitted when
    BOTH sides have real data; the magnitude is the size of the disagreement."""

    dimension: str
    stated: str
    revealed: str
    finding: str
    magnitude: float = Field(ge=0.0, le=1.0)
    provenance: Provenance
    confidence: float = Field(ge=0.0, le=1.0)


class GapUncomputable(BaseModel):
    """A dimension we could not compare, naming WHICH side was missing. A gap analysis
    that silently omits the dimensions it lacked data for reads as agreement."""

    dimension: str
    missing: str
    reason: str


class GapReport(BaseModel):
    findings: list[GapFinding] = Field(default_factory=list)
    uncomputable: list[GapUncomputable] = Field(default_factory=list)
    agreements: list[str] = Field(default_factory=list)
    computed_at: datetime
